### ISSUE TITLE
VALUES clause fails to parse in spark3

### ISSUE THREAD

**Issue by mcannamela** [sqlfluff/sqlfluff#2152]
## Expected Behaviour
`VALUES` clauses should parse, and correctly identify the subsequent literals. 

## Observed Behaviour
Parsing succeeds with one record and no alias, but treats it as a function call: `...VALUES (...);`
Parsing fails when using an alias: `...VALUES (...) as (...);`
Parsing fails when using multiple records: `...VALUES (...), (...)...(...);`

## Steps to Reproduce
See this example fail CI [here](https://github.com/mcannamela/sqlfluff/actions/runs/1607132440).

`sqlfluff parse` the following:
```
select a, b from values (1, 2);
select a, b from values (1, 2) as (a,b);
select a, b from values (1, 2), (3, 4);
```

result:
```
[L:  1, P:  1]      |file:
[L:  1, P:  1]      |    base:
[L:  1, P:  1]      |        select_statement:
[L:  1, P:  1]      |            select_clause:
[L:  1, P:  1]      |                keyword:                                      'select'
[L:  1, P:  7]      |                [META] indent:
[L:  1, P:  7]      |                whitespace:                                   ' '
[L:  1, P:  8]      |                select_clause_element:
[L:  1, P:  8]      |                    column_reference:
[L:  1, P:  8]      |                        identifier:                           'a'
[L:  1, P:  9]      |                comma:                                        ','
[L:  1, P: 10]      |                whitespace:                                   ' '
[L:  1, P: 11]      |                select_clause_element:
[L:  1, P: 11]      |                    column_reference:
[L:  1, P: 11]      |                        identifier:                           'b'
[L:  1, P: 12]      |            whitespace:                                       ' '
[L:  1, P: 13]      |            [META] dedent:
[L:  1, P: 13]      |            from_clause:
[L:  1, P: 13]      |                keyword:                                      'from'
[L:  1, P: 17]      |                whitespace:                                   ' '
[L:  1, P: 18]      |                from_expression:
[L:  1, P: 18]      |                    [META] indent:
[L:  1, P: 18]      |                    from_expression_element:
[L:  1, P: 18]      |                        table_expression:
[L:  1, P: 18]      |                            function:
[L:  1, P: 18]      |                                function_name:
[L:  1, P: 18]      |                                    function_name_identifier:  'values'
[L:  1, P: 24]      |                                whitespace:                   ' '
[L:  1, P: 25]      |                                bracketed:
[L:  1, P: 25]      |                                    start_bracket:            '('
[L:  1, P: 26]      |                                    [META] indent:
[L:  1, P: 26]      |                                    expression:
[L:  1, P: 26]      |                                        literal:              '1'
[L:  1, P: 27]      |                                    comma:                    ','
[L:  1, P: 28]      |                                    whitespace:               ' '
[L:  1, P: 29]      |                                    expression:
[L:  1, P: 29]      |                                        literal:              '2'
[L:  1, P: 30]      |                                    [META] dedent:
[L:  1, P: 30]      |                                    end_bracket:              ')'
[L:  1, P: 31]      |                    [META] dedent:
[L:  1, P: 31]      |    statement_terminator:                                     ';'
[L:  1, P: 32]      |    newline:                                                  '\n'
[L:  2, P:  1]      |    base:
[L:  2, P:  1]      |        select_statement:
[L:  2, P:  1]      |            select_clause:
[L:  2, P:  1]      |                keyword:                                      'select'
[L:  2, P:  7]      |                [META] indent:
[L:  2, P:  7]      |                whitespace:                                   ' '
[L:  2, P:  8]      |                select_clause_element:
[L:  2, P:  8]      |                    column_reference:
[L:  2, P:  8]      |                        identifier:                           'a'
[L:  2, P:  9]      |                comma:                                        ','
[L:  2, P: 10]      |                whitespace:                                   ' '
[L:  2, P: 11]      |                select_clause_element:
[L:  2, P: 11]      |                    column_reference:
[L:  2, P: 11]      |                        identifier:                           'b'
[L:  2, P: 12]      |            whitespace:                                       ' '
[L:  2, P: 13]      |            [META] dedent:
[L:  2, P: 13]      |            from_clause:
[L:  2, P: 13]      |                keyword:                                      'from'
[L:  2, P: 17]      |                whitespace:                                   ' '
[L:  2, P: 18]      |                from_expression:
[L:  2, P: 18]      |                    [META] indent:
[L:  2, P: 18]      |                    from_expression_element:
[L:  2, P: 18]      |                        table_expression:
[L:  2, P: 18]      |                            function:
[L:  2, P: 18]      |                                function_name:
[L:  2, P: 18]      |                                    function_name_identifier:  'values'
[L:  2, P: 24]      |                                whitespace:                   ' '
[L:  2, P: 25]      |                                bracketed:
[L:  2, P: 25]      |                                    start_bracket:            '('
[L:  2, P: 26]      |                                    [META] indent:
[L:  2, P: 26]      |                                    expression:
[L:  2, P: 26]      |                                        literal:              '1'
[L:  2, P: 27]      |                                    comma:                    ','
[L:  2, P: 28]      |                                    whitespace:               ' '
[L:  2, P: 29]      |                                    expression:
[L:  2, P: 29]      |                                        literal:              '2'
[L:  2, P: 30]      |                                    [META] dedent:
[L:  2, P: 30]      |                                    end_bracket:              ')'
[L:  2, P: 31]      |                    whitespace:                               ' '
[L:  2, P: 32]      |                    [META] dedent:
[L:  2, P: 32]      |                unparsable:                                   !! Expected: 'Nothing...'
[L:  2, P: 32]      |                    raw:                                      'as'
[L:  2, P: 34]      |                    whitespace:                               ' '
[L:  2, P: 35]      |                    bracketed:
[L:  2, P: 35]      |                        start_bracket:                        '('
[L:  2, P: 36]      |                        raw:                                  'a'
[L:  2, P: 37]      |                        comma:                                ','
[L:  2, P: 38]      |                        raw:                                  'b'
[L:  2, P: 39]      |                        end_bracket:                          ')'
[L:  2, P: 40]      |    statement_terminator:                                     ';'
[L:  2, P: 41]      |    newline:                                                  '\n'
[L:  3, P:  1]      |    base:
[L:  3, P:  1]      |        select_statement:
[L:  3, P:  1]      |            select_clause:
[L:  3, P:  1]      |                keyword:                                      'select'
[L:  3, P:  7]      |                [META] indent:
[L:  3, P:  7]      |                whitespace:                                   ' '
[L:  3, P:  8]      |                select_clause_element:
[L:  3, P:  8]      |                    column_reference:
[L:  3, P:  8]      |                        identifier:                           'a'
[L:  3, P:  9]      |                comma:                                        ','
[L:  3, P: 10]      |                whitespace:                                   ' '
[L:  3, P: 11]      |                select_clause_element:
[L:  3, P: 11]      |                    column_reference:
[L:  3, P: 11]      |                        identifier:                           'b'
[L:  3, P: 12]      |            whitespace:                                       ' '
[L:  3, P: 13]      |            [META] dedent:
[L:  3, P: 13]      |            from_clause:
[L:  3, P: 13]      |                unparsable:                                   !! Expected: 'FromClauseSegment'
[L:  3, P: 13]      |                    keyword:                                  'from'
[L:  3, P: 17]      |                    whitespace:                               ' '
[L:  3, P: 18]      |                    raw:                                      'values'
[L:  3, P: 24]      |                    whitespace:                               ' '
[L:  3, P: 25]      |                    bracketed:
[L:  3, P: 25]      |                        start_bracket:                        '('
[L:  3, P: 26]      |                        [META] indent:
[L:  3, P: 26]      |                        expression:
[L:  3, P: 26]      |                            literal:                          '1'
[L:  3, P: 27]      |                        comma:                                ','
[L:  3, P: 28]      |                        whitespace:                           ' '
[L:  3, P: 29]      |                        expression:
[L:  3, P: 29]      |                            literal:                          '2'
[L:  3, P: 30]      |                        [META] dedent:
[L:  3, P: 30]      |                        end_bracket:                          ')'
[L:  3, P: 31]      |                    comma:                                    ','
[L:  3, P: 32]      |                    whitespace:                               ' '
[L:  3, P: 33]      |                    bracketed:
[L:  3, P: 33]      |                        start_bracket:                        '('
[L:  3, P: 34]      |                        raw:                                  '3'
[L:  3, P: 35]      |                        comma:                                ','
[L:  3, P: 36]      |                        whitespace:                           ' '
[L:  3, P: 37]      |                        raw:                                  '4'
[L:  3, P: 38]      |                        end_bracket:                          ')'
[L:  3, P: 39]      |    statement_terminator:                                     ';'
[L:  3, P: 40]      |    newline:                                                  '\n'

==== parsing violations ====
L:   2 | P:  32 |  PRS | Line 2, Position 32: Found unparsable section: 'as (a,b)'
L:   3 | P:  13 |  PRS | Line 3, Position 13: Found unparsable section: 'from values (1,
                       | 2), (3, 4)'

```

## Dialect
spark3

## Version
sqlfluff, version 0.9.0
Python 3.8.12

## Configuration
```
[sqlfluff]
verbose = 0
nocolor = False
dialect = spark3
templater = jinja
rules = None

recurse = 0
output_line_length = 80
runaway_limit = 10
ignore_templated_areas = True
```

**Comment by mcannamela** [comment]
Hm, suspicious comment [here](https://github.com/sqlfluff/sqlfluff/blob/c540632c654d719b4f5507f6925b398716bcb42b/src/sqlfluff/dialects/dialect_ansi.py#L1852), though I confess I'm not sure I understand what it's getting at. 

But the definition of `ValuesClauseSegment` looks right to my untrained eye.

**Comment by mcannamela** [comment]
I have it partly working with changes [here](https://github.com/mcannamela/sqlfluff/blob/7607a1cc8c5255e1b6c8a46b711f5dcd129a5a48/src/sqlfluff/dialects/dialect_spark3.py#L787), with CI [here](https://github.com/mcannamela/sqlfluff/actions/runs/1612455382).

```
select a, b from values (1, 2);
```
parses correctly
```
select a, b from values (1, 2), (3, 4), (greatest(5, 6), least(7, 8));
```

fails, for reasons I don't understand.

**Comment by tunetheweb** [comment]
> I have it partly working with changes [here](https://github.com/mcannamela/sqlfluff/blob/7607a1cc8c5255e1b6c8a46b711f5dcd129a5a48/src/sqlfluff/dialects/dialect_spark3.py#L787), with CI [here](https://github.com/mcannamela/sqlfluff/actions/runs/1612455382).
> 
> ```
> select a, b from values (1, 2);
> ```
> 
> parses correctly
> 
> ```
> select a, b from values (1, 2), (3, 4), (greatest(5, 6), least(7, 8));
> ```
> 
> fails, for reasons I don't understand.

This is another case of `Delimited` being greedy. SQLfluff see's this as three `from` clauses:

```sql
select a, b
from
    values (1, 2),
    (3, 4),
    (greatest(5, 6), least(7, 8));
```

Rather than all being part of one `VALUES` statement. Then it fails to parse because the last two aren't valid `FROM` statements. Then it doesn't try again to see if there's an alternative way of Delimited-ing the SQL.

Not sure how to get around that to. be honest without a fundamental change to how `Delimited` works....

**Comment by BI-MarcB** [comment]
The same parsing error appears with Snowflake as well when using a VALUES list in the FROM clause: https://docs.snowflake.com/en/sql-reference/constructs/values.html

Is it possible to port this fix to the Snowflake dialect as well or should this rather be raised in a separate issue?
